go/batch-submitter/batch-tx/encoding.go

```diff
@@ -0,0 +1,208 @@
+//  - num_contexts:                   3 bytes
+//    - num_contexts * batch_context: num_contexts * 16 bytes
+//  - [num txs ommitted]
+//    - tx_len:                       3 bytes
+//    - tx_bytes:                     tx_len bytes
+func (p *AppendSequencerBatchParams) Write(w *bytes.Buffer) error {
+	writeUint64(w, p.ShouldStartAtElement, 5)
+	writeUint64(w, p.TotalElementsToAppend, 3)
+
+	// Write number of contexts followed by each fixed-size BatchContext.
+	writeUint64(w, uint64(len(p.Contexts)), 3)
+	for _, context := range p.Contexts {
+		context.Write(w)
+	}
+
+	// Write each length-prefixed tx.
+	var txBuf bytes.Buffer
+	for _, tx := range p.Txs {
+		txBuf.Reset()
+
+		if err := tx.EncodeRLP(&txBuf); err != nil {
+			return err
+		}
+
+		writeUint64(w, uint64(txBuf.Len()), 3)
+		_, _ = w.Write(txBuf.Bytes()) // can't fail for bytes.Buffer
+	}
+
+	return nil
+}
+
+// Serialize performs the same encoding as Write, but returns the resulting
+// bytes slice.
+func (p *AppendSequencerBatchParams) Serialize() ([]byte, error) {
+	var buf bytes.Buffer
+	if err := p.Write(&buf); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// Read decodes the AppendSequencerBatchParams from a bytes stream. If the byte
+// stream does not terminate cleanly with an EOF while reading a tx_len, this
+// method will return an error. Otherwise, the stream will be parsed according
+// to the following format:
+//  - should_start_at_element:        5 bytes
+//  - total_elements_to_append:       3 bytes
+//  - num_contexts:                   3 bytes
+//    - num_contexts * batch_context: num_contexts * 16 bytes
+//  - [num txs ommitted]
+//    - tx_len:                       3 bytes
+//    - tx_bytes:                     tx_len bytes
+func (p *AppendSequencerBatchParams) Read(r io.Reader) error {
+	if err := readUint64(r, &p.ShouldStartAtElement, 5); err != nil {
+		return err
+	}
+	if err := readUint64(r, &p.TotalElementsToAppend, 3); err != nil {
+		return err
+	}
+
+	// Read number of contexts and deserialize each one.
+	var numContexts uint64
+	if err := readUint64(r, &numContexts, 3); err != nil {
+		return err
+	}
+
+	for i := uint64(0); i < numContexts; i++ {
+		var batchContext BatchContext
+		if err := batchContext.Read(r); err != nil {
+			return err
+		}
+
+		p.Contexts = append(p.Contexts, batchContext)
+	}
+
+	// Deserialize any transactions. Since the number of txs is ommitted
+	// from the encoding, loop until the stream is consumed.
+	for {
+		var txLen uint64
+		err := readUint64(r, &txLen, 3)
+		// Getting an EOF when reading the txLen expected for a cleanly
+		// encoded object. Silece the error and return success.
+		if err == io.EOF {
+			return nil
+		} else if err != nil {
+			return err
+		}
+
+		tx := new(l2types.Transaction)
+		if err := tx.DecodeRLP(l2rlp.NewStream(r, txLen)); err != nil {
+			return err
+		}
+
+		p.Txs = append(p.Txs, tx)
+	}
+}
+
+// writeUint64 writes a the bottom `n` bytes of `val` to `w`.
```

Comment:
What do you think about adding overflow protection here?

Reply:
added 👍